opnode/rollup/derive/payload_attributes.go

```diff
@@ -178,43 +181,166 @@ func UserDeposits(height uint64, receipts []*types.Receipt) ([]*types.DepositTx,
 
 type BlockInput interface {
 	ReceiptHash
+	ParentHash() common.Hash
 	L1Info
 	MixDigest() common.Hash
+	Transactions() []*types.Transaction
+	Receipts() []*types.Receipt
 }
 
 // PayloadAttributes derives the pre-execution payload from the L1 block info and deposit receipts.
 // This is a pure function.
-func PayloadAttributes(block BlockInput, receipts []*types.Receipt) (*l2.PayloadAttributes, error) {
-	if !CheckReceipts(block, receipts) {
-		return nil, fmt.Errorf("receipts are not consistent with the block's receipts root: %s", block.ReceiptHash())
+func PayloadAttributes(config *rollup.Config, ontoL1 eth.BlockID, lastL2Time uint64, blocks []BlockInput) ([]*l2.PayloadAttributes, error) {
+	// check if we have the full sequencing window
+	if uint64(len(blocks)) != config.SeqWindowSize {
+		return nil, fmt.Errorf("expected %d blocks in sequencing window, got %d", config.SeqWindowSize, len(blocks))
+	}
+	// check if our inputs are consistent
+	for i, block := range blocks {
+		if block.NumberU64() != ontoL1.Number+1 {
+			return nil, fmt.Errorf("sanity check failed, sequencing window blocks are not consecutive (index %d)", i)
+		}
+		// sanity check the parent hash
+		if block.ParentHash() != ontoL1.Hash {
+			return nil, fmt.Errorf("sanity check failed, sequencing window blocks are not chained  (index %d)", i)
+		}
+		ontoL1 = eth.BlockID{Hash: block.Hash(), Number: block.NumberU64()}
+	}
+
+	// Retrieve the deposits of this epoch (all deposits from the first block)
+	deposits, err := DeriveDeposits(blocks[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive deposits: %v", err)
+	}
+
+	// All block times in L2 are config.BlockTime seconds apart. If L2 time moves faster than L1, use that.
+	// L1 is expected to have larger increments per block, L2 time will get back in sync eventually.
+	firstTimestamp := blocks[0].Time()
+	if firstTimestamp >= config.Genesis.L2Time { // if not, then lastL2Time will be larger
+		// round down, only create a trailing empty block if the L1 time is at least a full block span ahead of L2
+		firstTimestamp = firstTimestamp - (firstTimestamp-config.Genesis.L2Time)%config.BlockTime
 	}
+	if lastL2Time+config.BlockTime > firstTimestamp {
+		firstTimestamp = lastL2Time + config.BlockTime
+	}
+
+	// Sequencers may not use timestamps beyond the end of the sequencing window (with some configurable margin)
+	maxTimestamp := blocks[len(blocks)-1].Time() + config.MaxSequencerTimeDiff
```

Comment:
While it does seem sane, is there a particular reason why do we have this constraint? i.e. what's the bad thing that can happen if this condition is violated?